File: src/prometheus/lock.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");

const LOCK_FILE = "prometheus.lock";
const STATE_FILE = "state.json";

class Lock {
    constructor({ dataDir, logger } = {}) {
        if (!dataDir) throw new Error("dataDir is required");
        if (!logger) throw new Error("logger is required");
        this.dataDir = dataDir;
        this.logger = logger;
        this.lockPath = path.join(dataDir, LOCK_FILE);
        this.statePath = path.join(dataDir, STATE_FILE);
        fs.mkdirSync(dataDir, { recursive: true });
    }

    _atomicWrite(state) {
        state.ts = Date.now();
        const tmp = this.lockPath + ".tmp";
        fs.writeFileSync(tmp, JSON.stringify(state, null, 2));
        fs.renameSync(tmp, this.lockPath);
    }

    acquire(cycleNum) {
        const old = this._readLock();
        if (old && old.pid && old.dirty) {
            let alive = false;
            try {
                process.kill(old.pid, 0);
                alive = true;
            } catch (e) {
                alive = false;
            }
            if (alive) {
                throw new Error("another instance running (pid=" + old.pid + ")");
            }
            this.logger.log(
                "WARN",
                "stale lock from dead pid=" + old.pid + ", overwriting"
            );
        }
        const state = {
            pid: process.pid,
            dirty: true,
            cycle: cycleNum,
            bottomReached: old ? old.bottomReached : undefined,
            pending_media: [],
            ts: Date.now(),
        };
        this._atomicWrite(state);
        return state;
    }

    _readLock() {
        if (!fs.existsSync(this.lockPath)) return null;
        try {
            return JSON.parse(fs.readFileSync(this.lockPath, "utf8"));
        } catch (e) {
            return null;
        }
    }

    _writeLock(state) {
        this._atomicWrite(state);
    }

    addPendingMedia(url, file, expectedSize) {
        const state = this._readLock();
        if (!state) throw new Error("no lock held");
        if (!Array.isArray(state.pending_media)) state.pending_media = [];
        state.pending_media.push({ url, file, expected_size: expectedSize });
        this._writeLock(state);
    }

    removePendingMedia(url) {
        const state = this._readLock();
        if (!state) throw new Error("no lock held");
        if (!Array.isArray(state.pending_media)) return;
        state.pending_media = state.pending_media.filter((m) => m.url !== url);
        this._writeLock(state);
    }

    release() {
        const state = this._readLock();
        if (state) {
            state.dirty = false;
            state.pending_media = [];
            this._writeLock(state);
        }
    }

    setBottomReached(value) {
        const state = this._readLock();
        if (state && state.bottomReached === value) return;
        const next = state || {
            pid: process.pid,
            dirty: false,
            cycle: 0,
            pending_media: [],
        };
        next.bottomReached = value;
        this._atomicWrite(next);
    }

    readBottomReached() {
        const state = this._readLock();
        if (!state) return undefined;
        return state.bottomReached;
    }

    checkAndRecover() {
        const lockState = this._readLock();
        if (!lockState) return null;
        if (lockState.dirty) {
            return {
                crashed: true,
                bottomReached: lockState.bottomReached,
                pendingMedia: Array.isArray(lockState.pending_media)
                    ? lockState.pending_media
                    : [],
            };
        }
        return null;
    }
}

module.exports = { Lock };

File: src/web_scraper/lock.py
"""Process-level lock with stale-PID detection and crash recovery."""

from __future__ import annotations

import json
import os
import time
from pathlib import Path


class LockError(Exception):
    pass


class Lock:
    LOCK_FILE = "prometheus.lock"

    def __init__(self, data_dir: Path, logger=None):
        self.data_dir = Path(data_dir)
        self.data_dir.mkdir(parents=True, exist_ok=True)
        self.lock_path = self.data_dir / self.LOCK_FILE
        self.logger = logger
        self._held = False

    def _read(self) -> dict | None:
        if not self.lock_path.exists():
            return None
        try:
            return json.loads(self.lock_path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return None

    def _write(self, state: dict) -> None:
        tmp = self.lock_path.with_suffix(".tmp")
        tmp.write_text(json.dumps(state, indent=2), encoding="utf-8")
        tmp.replace(self.lock_path)

    @staticmethod
    def _pid_alive(pid: int) -> bool:
        try:
            os.kill(pid, 0)
            return True
        except (ProcessLookupError, PermissionError):
            return False
        except OSError:
            return False

    def acquire(self, cycle: int = 0) -> dict:
        old = self._read()
        if old and old.get("pid") and old.get("dirty"):
            pid = old["pid"]
            if self._pid_alive(pid):
                raise LockError(
                    f"another instance running (pid={pid})"
                )
            if self.logger:
                self.logger.warning(
                    "stale lock from dead pid=%s, overwriting", pid
                )

        state = {
            "pid": os.getpid(),
            "dirty": True,
            "cycle": cycle,
            "ts": int(time.time()),
        }
        self._write(state)
        self._held = True
        return state

    def release(self) -> None:
        if not self._held:
            return
        state = self._read()
        if state:
            state["dirty"] = False
            self._write(state)
        self._held = False

    def check_and_recover(self) -> dict | None:
        old = self._read()
        if not old or not old.get("dirty"):
            return None
        pid = old.get("pid", 0)
        if pid and self._pid_alive(pid):
            return None
        return {
            "crashed": True,
            "cycle": old.get("cycle"),
            "ts": old.get("ts"),
        }
